forum/views.py

```diff
@@ -1,3 +1,74 @@
-from django.shortcuts import render
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.views.generic.detail import DetailView
+from django.views.generic.list import ListView
+from django.views import View
+from django.shortcuts import get_object_or_404, redirect
 
-# Create your views here.
+from .models import (
+    Forum,
+    Post,
+    Comment,
+    CommentVote,
+    UPVOTE,
+    DOWNVOTE
+)
+
+
+class ForumDetailView(DetailView):
+    model = Forum
+
+
+class ForumListView(ListView):
+    model = Forum
+
+
+class PostDetailView(DetailView):
+    model = Post
+    pk_url_kwarg = 'post_id'
+
+
+class CommentDetailView(DetailView):
+    model = Comment
+    pk_url_kwarg = 'comment_id'
+
+
+class CommentUpvoteView(LoginRequiredMixin, View):
+    def post(self, request, *args, **kwargs):
+        comment = get_object_or_404(Comment, id=self.kwargs.get('pk'))
+
+        try:
+            comment_vote = CommentVote.objects.get(
```

Comment:
When you upvote, you have to check if a vote by that User already exists. If so, set it to `UPVOTE`. Otherwise, create an entirely new vote.

(do the opposite for downvotes)